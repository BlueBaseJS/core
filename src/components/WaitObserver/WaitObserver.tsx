import React from 'react';

export interface WaitObserverChildrenProps {

	/** A flag that tells if a timeout has occured */
	timedOut: boolean,

	/** A function the resets timers */
	retry: () => void
}

export interface WaitObserverProps {

	/** Delay before rendering a component */
	delay?: number,

	/** Timeout duration */
	timeout?: number,

	/**
	 * A callback function executed when a timeout occurs.
	 */
	onTimeout?: () => void;

	/**
	 * A callback function executed when retry method is called from the child component.
	 */
	onRetry?: () => void;

	children: ((props: WaitObserverChildrenProps) => React.ReactNode);
}

interface WaitObserverState {

	/**
	 * Flag that tells if we have waited more time than specified in the delay prop
	 */
	readonly pastDelay: boolean,

	/**
	 * Flag that tells if waiting has timedout
	 */
  readonly timedOut: boolean,
}

/**
 * ‚è∞ **WaitObserver Component**
 *
 * This component is used to do the following:
 *
 * - WaitObserver a certain period of time before rendering a component
 * - Show timeout state, if the component is visible for a certain time period
 *
 * A use case for this can be to show a loading state after waiting a certain period
 * of time for data to load, and if the loading takes too long, show a timeout state.
 */
export class WaitObserver extends React.PureComponent<WaitObserverProps> {

	public static defaultProps: Partial<WaitObserverProps> = {
		delay: 200,
		onTimeout: () => { return; },
		onRetry: () => { return; },
	};

	readonly state: WaitObserverState = {
		pastDelay: false,
    timedOut: false,
	};

	private _delay?: NodeJS.Timer;
	private _timeout?: NodeJS.Timer;

	componentWillMount() {
		this.init();
	}

	componentWillUnmount() {
		this.clearTimeouts();
	}

	render() {
		if (this.state.pastDelay) {
			return this.props.children({
				timedOut: this.state.timedOut,
				retry: this.retry
			});
		} else {
			return null;
		}
	}

	private init() {
		if (typeof this.props.delay === 'number') {
			if (this.props.delay === 0) {
				this.setState({ pastDelay: true });
			} else {
				this._delay = setTimeout(() => {
					this.setState({ pastDelay: true });
				}, this.props.delay);
			}
		}

		if (typeof this.props.timeout === 'number') {
			this._timeout = setTimeout(() => {
				this.props.onTimeout && this.props.onTimeout();
				this.setState({ timedOut: true });
			}, this.props.timeout);
		}
	}

	private clearTimeouts() {
		this._delay && clearTimeout(this._delay);
		this._timeout && clearTimeout(this._timeout);
	}

	private retry = () => {
		this.props.onRetry && this.props.onRetry();
		this.setState({ timedOut: false });
		this.init();
	}
}
