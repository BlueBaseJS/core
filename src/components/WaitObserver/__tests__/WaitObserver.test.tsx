import { Text, View } from 'react-native';
import React from 'react';
// import TestRenderer from 'react-test-renderer';
import { WaitObserver } from '../WaitObserver';
import { mount } from 'enzyme';

jest.useFakeTimers();

const LoadingState = (props: any) => {

	const { timedOut } = props;

	let title = 'Loading now...';

	if (timedOut === true) {
		title = 'This is taking longer than usual';

		// if (retry) {
		// 	props.actionTitle = 'Retry';
		// 	props.actionOnPress = retry;
		// }
	}

	return (
		<View>
			<Text testID="title">{title}</Text>
		</View>
	);
};

describe.only('WaitObserver', () => {

	test(`should change states based on custom delay & timeout`, () => {

		const onTimeout = jest.fn();
		const onRetry = jest.fn();

		const component = mount(
			<WaitObserver
				delay={1000}
				timeout={3000}
				onTimeout={onTimeout}
				onRetry={onRetry}
				children={(props: any) => <LoadingState {...props} />}
			/>
		);

		// Should render null
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(0);
		expect(onRetry.mock.calls.length).toBe(0);
		expect(component.find('WaitObserver').text()).toBeNull();
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: false,
			timedOut: false,
		});

		// Fast-forward
		jest.advanceTimersByTime(1000);
		component.update();

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(0);
		expect(onRetry.mock.calls.length).toBe(0);
		expect(component.find('WaitObserver').text()).toBe('Loading now...');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: false,
		});

		// Fast-forward
		jest.advanceTimersByTime(2000);
		component.update();

		// Should render timeout message
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(1);
		expect(onRetry.mock.calls.length).toBe(0);
		expect(component.find('WaitObserver').text()).toBe('This is taking longer than usual');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: true,
		});

		// Retry
		const instance = component.find('WaitObserver').instance();
		(instance as any).retry();
		component.update();

		//////// Start the process again

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(1);
		expect(onRetry.mock.calls.length).toBe(1);
		expect(component.find('WaitObserver').text()).toBe('Loading now...');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: false,
		});

		// Fast-forward
		jest.advanceTimersByTime(3000);
		component.update();

		// Should render timeout message
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(2);
		expect(onRetry.mock.calls.length).toBe(1);
		expect(component.find('WaitObserver').text()).toBe('This is taking longer than usual');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: true,
		});
	});

	test(`should change states based on custom delay & timeouts`, () => {

		const onTimeout = jest.fn();

		const Comp = WaitObserver as any;

		const component = mount(
			<Comp
				delay={false}
				timeout={false}
				onTimeout={onTimeout}
				children={(props: any) => <LoadingState {...props} />}
			/>
		);

		// Should render null
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(0);
		expect(component.find('WaitObserver').text()).toBeNull();
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: false,
			timedOut: false,
		});

		// Fast-forward
		jest.advanceTimersByTime(1000);
		component.update();

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(0);
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: false,
			timedOut: false,
		});

		// Fast-forward
		component.unmount();

		// Should render timeout message
		expect(component).toMatchSnapshot();
		expect(component.getElements()).toHaveLength(0);

	});

	test(`should not change states based on timeouts, becuase it gets unmounted`, () => {

		const onTimeout = jest.fn();

		const component = mount(
			<WaitObserver
				delay={1000}
				timeout={3000}
				onTimeout={onTimeout}
				children={(props: any) => <LoadingState {...props} />}
			/>
		);

		// Should render null
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(0);
		expect(component.find('WaitObserver').text()).toBeNull();
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: false,
			timedOut: false,
		});

		// unmount
		component.unmount();

		// Fast-forward
		jest.advanceTimersByTime(1000);
		component.update();

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(onTimeout.mock.calls.length).toBe(0);
		expect(component.getElements()).toHaveLength(0);

		// Fast-forward
		jest.advanceTimersByTime(2000);
		component.update();

		// Should render timeout message
		expect(component).toMatchSnapshot();
		expect(component.getElements()).toHaveLength(0);

	});

	test(`should show loading state if delay is 0`, () => {

		const component = mount(
			<WaitObserver
				delay={0}
				timeout={2000}
				onRetry={null as any}
				onTimeout={null as any}
				children={<LoadingState />}
			/>
		);

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(component.find('WaitObserver').text()).toBe('Loading now...');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: false,
		});

		// Fast-forward
		jest.advanceTimersByTime(2000);
		component.update();

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(component.find('WaitObserver').text()).toBe('Loading now...');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: true,
		});

		// Retry
		const instance = component.find('WaitObserver').instance();
		(instance as any).retry();
		component.update();

		// Should render Loading message
		expect(component).toMatchSnapshot();
		expect(component.find('WaitObserver').text()).toBe('Loading now...');
		expect(component.find('WaitObserver').state()).toMatchObject({
			pastDelay: true,
			timedOut: false,
		});

	});

	test(`default callback functions should return undefined`, () => {

		const component = mount(
			<WaitObserver
				children={<LoadingState />}
			/>
		);

		const onRetry: any = component.find('WaitObserver').prop('onRetry');
		const onTimeout: any = component.find('WaitObserver').prop('onTimeout');

		expect(onRetry()).toBe(undefined);
		expect(onTimeout()).toBe(undefined);

	});

});
