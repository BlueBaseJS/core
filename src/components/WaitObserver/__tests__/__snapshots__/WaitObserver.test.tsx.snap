// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WaitObserver should change states based on custom delay & timeout 1`] = `
<WaitObserver
  delay={1000}
  onRetry={[MockFunction]}
  onTimeout={[MockFunction]}
  timeout={3000}
/>
`;

exports[`WaitObserver should change states based on custom delay & timeout 2`] = `
<WaitObserver
  delay={1000}
  onRetry={[MockFunction]}
  onTimeout={[MockFunction]}
  timeout={3000}
>
  <LoadingState
    retry={[Function]}
    timedOut={false}
  >
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            Loading now...
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;

exports[`WaitObserver should change states based on custom delay & timeout 3`] = `
<WaitObserver
  delay={1000}
  onRetry={[MockFunction]}
  onTimeout={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  timeout={3000}
>
  <LoadingState
    retry={[Function]}
    timedOut={true}
  >
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            This is taking longer than usual
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;

exports[`WaitObserver should change states based on custom delay & timeout 4`] = `
<WaitObserver
  delay={1000}
  onRetry={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  onTimeout={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  timeout={3000}
>
  <LoadingState
    retry={[Function]}
    timedOut={false}
  >
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            Loading now...
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;

exports[`WaitObserver should change states based on custom delay & timeout 5`] = `
<WaitObserver
  delay={1000}
  onRetry={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  onTimeout={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  timeout={3000}
>
  <LoadingState
    retry={[Function]}
    timedOut={true}
  >
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            This is taking longer than usual
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;

exports[`WaitObserver should change states based on custom delay & timeouts 1`] = `
<WaitObserver
  delay={false}
  onRetry={[Function]}
  onTimeout={[MockFunction]}
  timeout={false}
/>
`;

exports[`WaitObserver should change states based on custom delay & timeouts 2`] = `
<WaitObserver
  delay={false}
  onRetry={[Function]}
  onTimeout={[MockFunction]}
  timeout={false}
/>
`;

exports[`WaitObserver should change states based on custom delay & timeouts 3`] = `null`;

exports[`WaitObserver should not change states based on timeouts, becuase it gets unmounted 1`] = `
<WaitObserver
  delay={1000}
  onRetry={[Function]}
  onTimeout={[MockFunction]}
  timeout={3000}
/>
`;

exports[`WaitObserver should not change states based on timeouts, becuase it gets unmounted 2`] = `null`;

exports[`WaitObserver should not change states based on timeouts, becuase it gets unmounted 3`] = `null`;

exports[`WaitObserver should show loading state if delay is 0 1`] = `
<WaitObserver
  delay={0}
  onRetry={null}
  onTimeout={null}
  timeout={2000}
>
  <LoadingState>
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            Loading now...
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;

exports[`WaitObserver should show loading state if delay is 0 2`] = `
<WaitObserver
  delay={0}
  onRetry={null}
  onTimeout={null}
  timeout={2000}
>
  <LoadingState>
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            Loading now...
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;

exports[`WaitObserver should show loading state if delay is 0 3`] = `
<WaitObserver
  delay={0}
  onRetry={null}
  onTimeout={null}
  timeout={2000}
>
  <LoadingState>
    <View>
      <View>
        <Text
          testID="title"
        >
          <Text
            testID="title"
          >
            Loading now...
          </Text>
        </Text>
      </View>
    </View>
  </LoadingState>
</WaitObserver>
`;
